pkg/template/controller/templateinstance_controller.go

```diff
@@ -132,38 +140,130 @@ func (c *TemplateInstanceController) sync(key string) error {
+			})
+		}
+	}
 
-	if instantiateErr == nil {
-		templateInstance.Status.Conditions = append(templateInstance.Status.Conditions, templateapi.TemplateInstanceCondition{
-			Type:               templateapi.TemplateInstanceReady,
-			Status:             kapi.ConditionTrue,
-			LastTransitionTime: metav1.Now(),
-			Reason:             "Created",
-		})
+	_, err = c.templateclient.TemplateInstances(templateInstance.Namespace).UpdateStatus(templateInstance)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("TemplateInstance status update failed: %v", err))
+		return err
+	}
 
+	if templateInstance.HasCondition(templateapi.TemplateInstanceWaiting, kapi.ConditionTrue) {
+		c.enqueueAfter(templateInstance, c.readinessLimiter.When(key))
 	} else {
-		templateInstance.Status.Conditions = append(templateInstance.Status.Conditions, templateapi.TemplateInstanceCondition{
-			Type:               templateapi.TemplateInstanceInstantiateFailure,
-			Status:             kapi.ConditionTrue,
-			LastTransitionTime: metav1.Now(),
-			Reason:             "Failed",
-			Message:            instantiateErr.Error(),
+		c.readinessLimiter.Forget(key)
+	}
+
+	return nil
+}
+
+func (c *TemplateInstanceController) checkReadiness(templateInstance *templateapi.TemplateInstance) error {
+	u := &user.DefaultInfo{Name: templateInstance.Spec.Requester.Username}
+
+	allReady := true
+	for _, object := range templateInstance.Status.Objects {
+		if !canCheckReadiness(object.Ref) {
+			continue
+		}
+
+		mapping, err := c.restmapper.RESTMapping(object.Ref.GroupVersionKind().GroupKind())
+		if err != nil {
+			return err
+		}
+
+		if err := util.Authorize(c.kc.Authorization().SubjectAccessReviews(), u, &authorization.ResourceAttributes{
+			Namespace: object.Ref.Namespace,
+			Verb:      "get",
+			Group:     object.Ref.GroupVersionKind().Group,
+			Resource:  mapping.Resource,
+			Name:      object.Ref.Name,
+		}); err != nil {
+			return err
+		}
+
+		cli, err := cmd.ClientMapperFromConfig(c.config).ClientForMapping(mapping)
+		if err != nil {
+			return err
+		}
+
+		obj, err := cli.Get().Resource(mapping.Resource).NamespaceIfScoped(object.Ref.Namespace, mapping.Scope.Name() == meta.RESTScopeNameNamespace).Name(object.Ref.Name).Do().Get()
+		if err != nil {
+			return err
+		}
+
+		meta, err := meta.Accessor(obj)
+		if err != nil {
+			return err
+		}
+
+		if meta.GetUID() != object.Ref.UID {
+			return kerrors.NewNotFound(schema.GroupResource{Group: mapping.GroupVersionKind.Group, Resource: mapping.Resource}, object.Ref.Name)
+		}
+
+		if strings.ToLower(meta.GetAnnotations()[templateapi.WaitForReadyAnnotation]) != "true" {
+			continue
+		}
+
+		ready, failed := checkReadiness(object.Ref, obj)
+		if failed {
+			return fmt.Errorf("readiness failed on %s %s/%s", object.Ref.Kind, object.Ref.Namespace, object.Ref.Name)
+		}
+		if !ready {
+			allReady = false
+			break
+		}
+	}
+
+	if allReady {
+		templateInstance.DeleteCondition(templateapi.TemplateInstanceWaiting)
```

Comment:
also wondering if checkReadiness should really be doing this, might be better for it just to return true/false and let the caller decide what to do w/ that information?  Updating the template instance feels like a side-effect here.